### ISSUE TITLE
List support

### ISSUE THREAD

**Issue by RickMoynihan** [yetanalytics/flint#32]
I don't think this is currently possible to do in flint, but it would be nice if it were possible to generate SPARQL like this:

```sparql
SELECT DISTINCT ?s ?score
WHERE {
  ?s ?p ?l.
  (?l ?score) <tag:stardog:api:property:textMatch> 'mac'.
}
```

and this:

```sparql
SELECT DISTINCT ?s ?score
WHERE {
  ?s ?p ?l.
  (?l ?score) <tag:stardog:api:property:textMatch> ('mac' 0.5 10).
}
```

In particular to access the [full-text search features of stardog](https://docs.stardog.com/query-stardog/full-text-search) which leverages this relatively rare (but valid) syntax.

**Comment by kelvinqian00** [comment]
I considered it at the beginning of Flint's development, but I decided against it because it would've introduced recursive tendencies that would've _massively_ complicated the grammar, all for a (as you said) relatively rare feature. It does get annoying when you want to build lists, but at the time I thought it was worth the tradeoff.

**Comment by RickMoynihan** [comment]
Yeah, I totally understand why you chose to omit this and focus on the more common features, as lists aren't widely used. 

Presumably though you mainly wanted to avoid it because it's uncommon and a bunch of extra complexity/work, rather than a desire to avoid recursion in the grammar itself, as expressions and subqueries are already recursive:

```clojure
'{:select [[(+ 1 (+ 2 (+ 3 (+ 4 5) 6 7 8 9))) ?sum]] :where []}
'{:select :* :where [[:where {:select :* :where [[:where {:select :* :where [[?s ?p ?o]]}]]}]]}
```

Regardless if you feel this isn't worth the complexity, which I totally understand; I wonder if there isn't already a mechanism to build queries like this whether you'd consider adding a special `unsafe-fragment` form, that can occur at any point in the grammar.  It would just splice a raw string into the SPARQL at that point; e.g.

 `{:select :* :where [[?s ?p ?l] #flint/unsafe-fragment "(?l ?score) <tag:stardog:api:property:textMatch> ('mac' 0.5 10)"]}

**Comment by kelvinqian00** [comment]
> Presumably though you mainly wanted to avoid it because it's uncommon and a bunch of extra complexity/work

Yeah of course, I meant _extra_ recursion when I said that.

> I wonder if there isn't already a mechanism to build queries like this, whether you'd consider adding a special unsafe-fragment form

That's an interesting idea, but I feel like it violates the philosophy of Flint where "EDN <-> SPARQL," plus (as you imply with the name) it can gunk up a whole bunch of internal mechanisms within Flint. If I do implement lists, I'd rather do so cleanly, either following the full spec or in a limited form that does not allow for recursion within lists.

**Comment by kelvinqian00** [comment]
@RickMoynihan Support for list and blank node collection syntax has been implemented in v0.3.0, in #36
